opt: plan row-level BEFORE triggers for cascading mutations

DrewKimball opened the pull request:
#### opt: mechanical changes for BEFORE triggers on cascades

This commit consists of the following changes to prepare for adding
support for row-level BEFORE triggers to cascading mutations:
* Added `unsafe_allow_triggers_modifying_cascades` which will determine
  whether a trigger can modify the rows mutated by a cascade, default off.
* Added a "cascade" argument to `buildRowLevelBeforeTriggers` to provide
  the context when building a row-level BEFORE trigger.
* Refactoring the logic for invoking `crdb_internal.plpgsql_raise` to be
  re-used for triggers.

Informs #132971

Release note: None

#### opt: plan row-level BEFORE triggers for cascading mutations

This commit adds logic to invoke the functions for row-level BEFORE
triggers that are fired by a cascade's mutation. This is mostly
straightforward, except for the fact that a row-level BEFORE trigger
can actually modify or filter rows-to-be-mutated. In Postgres, it's
possible to cause constraint violations via this behavior. This commit
adds a runtime check to ensure that trigger functions do not modify
cascading updates/deletes. The check is gated behind
`unsafe_allow_triggers_modifying_cascades`, default off.

Fixes #132971

Release note (sql change): Cascades can now fire row-level BEFORE triggers.
By default, attempting to modify or eliminate the cascading update/delete
will result in a `Triggered Data Change Violation` error. Users that wish
to do this anyway can set `unsafe_allow_triggers_modifying_cascades`.
Note that doing so could result in constraint violations, similar to Postgres.

#### sql: fix queuing behavior for checks and triggers

This commit fixes a bug in `PlanAndRunPostQueries` that could cause
check queries to be run more than once. This would cause problems with
unclosed resources after query completion. This commit also fixes a
minor bug that could cause triggers queued by triggers to be run
before newly-queued checks and cascades, instead of after. The following
commit includes a regression test.

Fixes #133792

Release note: None

#### opt: don't add spurious check queries after BEFORE trigger

This commit fixes a bug that could cause spurious constraint violation
errors when a BEFORE trigger was planned on a cascaded mutation. Since we
don't allow triggers to modify the rows of a cascaded mutation, it is safe
to avoid the extra check entirely. This commit also adds a diamond-pattern
cascade test, with triggers on each table.

Fixes #133784

Release note: None

DrewKimball committed:
opt: mechanical changes for BEFORE triggers on cascades

This commit consists of the following changes to prepare for adding
support for row-level BEFORE triggers to cascading mutations:
* Added `unsafe_allow_triggers_modifying_cascades` which will determine
  whether a trigger can modify the rows mutated by a cascade, default off.
* Added a "cascade" argument to `buildRowLevelBeforeTriggers` to provide
  the context when building a row-level BEFORE trigger.
* Refactoring the logic for invoking `crdb_internal.plpgsql_raise` to be
  re-used for triggers.

Informs #132971

Release note: None

DrewKimball committed:
opt: plan row-level BEFORE triggers for cascading mutations

This commit adds logic to invoke the functions for row-level BEFORE
triggers that are fired by a cascade's mutation. This is mostly
straightforward, except for the fact that a row-level BEFORE trigger
can actually modify or filter rows-to-be-mutated. In Postgres, it's
possible to cause constraint violations via this behavior. This commit
adds a runtime check to ensure that trigger functions do not modify
cascading updates/deletes. The check is gated behind
`unsafe_allow_triggers_modifying_cascades`, default off.

Fixes #132971

Release note (sql change): Cascades can now fire row-level BEFORE triggers.
By default, attempting to modify or eliminate the cascading update/delete
will result in a `Triggered Data Change Violation` error. Users that wish
to do this anyway can set `unsafe_allow_triggers_modifying_cascades`.
Note that doing so could result in constraint violations, similar to Postgres.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/133220)
<!-- Reviewable:end -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!!

After a BEFORE trigger modifies a row to be updated or deleted, do cascades use the new row? (And is there a test for that scenario?)

Is there a test showing a cascade diamond (where the same row is modified through two different cascades)? Would be nice to see that a trigger on the doubly-cascaded row fires twice with the correct old rows.

> Reviewed 13 of 13 files at r1, 26 of 26 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-:-OA4jn3j57uxc5BqERcr:bapcfku)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/opt/optbuilder/trigger.go` line 324 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OA4JzPj7glgzsP9prIc:-OA4JzPj7glgzsP9prId:b80it63) ([raw file](https://github.com/cockroachdb/cockroach/blob/da21da6f8291396cbd94e680637ce57d4acbe4e0/pkg/sql/opt/optbuilder/trigger.go#L324)):*
> ```Go
> 		memo.ScalarListExpr{
> 			f.ConstructWhen(
> 				f.ConstructIsNot(f.ConstructVariable(triggerFnColID), f.ConstructVariable(expectedColID)),
> ```

Will checking the entire row lead to false positives if the trigger changes non-FK columns? Would it be more accurate to only check the FK columns? (Could be a TODO if this sounds reasonable.)


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
>After a BEFORE trigger modifies a row to be updated or deleted, do cascades use the new row? (And is there a test for that scenario?)

Yes, and I'll add a test for that.

>Is there a test showing a cascade diamond (where the same row is modified through two different cascades)? Would be nice to see that a trigger on the doubly-cascaded row fires twice with the correct old rows.

That's a great idea. The test I wrote based on this suggestion actually caught a pair of bugs: #133784 and #133792. I'll open a PR to fix them + add the test.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-:-OAPFMJqCg3_-DhFG8sh:b-jm8ron)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/optbuilder/trigger.go` line 324 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OA4JzPj7glgzsP9prIc:-OAR1atuBfiTohfqMrFU:b-6ojsko) ([raw file](https://github.com/cockroachdb/cockroach/blob/da21da6f8291396cbd94e680637ce57d4acbe4e0/pkg/sql/opt/optbuilder/trigger.go#L324)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Will checking the entire row lead to false positives if the trigger changes non-FK columns? Would it be more accurate to only check the FK columns? (Could be a TODO if this sounds reasonable.)
</blockquote></details>

That does sound reasonable. I opened an issue #133787 and added a TODO that references it.


<!-- Sent from Reviewable.io -->

DrewKimball committed:
sql: fix queuing behavior for checks and triggers

This commit fixes a bug in `PlanAndRunPostQueries` that could cause
check queries to be run more than once. This would cause problems with
unclosed resources after query completion. This commit also fixes a
minor bug that could cause triggers queued by triggers to be run
before newly-queued checks and cascades, instead of after. The following
commit includes a regression test.

Fixes #133792

Release note: None

DrewKimball committed:
opt: don't add spurious check queries after BEFORE trigger

This commit fixes a bug that could cause spurious constraint violation
errors when a BEFORE trigger was planned on a cascaded mutation. Since we
don't allow triggers to modify the rows of a cascaded mutation, it is safe
to avoid the extra check entirely. This commit also adds a diamond-pattern
cascade test, with triggers on each table.

Fixes #133784

Release note: None

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 26 files at r2, 10 of 10 files at r6, 4 of 4 files at r7, 1 of 1 files at r8, 3 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-:-OAcTWvb2eAjjz4s6uen:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @michae2)

___
*[`-- commits` line 7 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OAcRKHjDGGWYh21HM-t):*
Can you explain more about the motivation for this setting? I don't see why it is necessary - Postgres allows triggers to modify cascading updates by default.

If we do want to keep it, I think the word "unsafe" is a bit too strong to use in this case - it sounds like it can lead to unexpected or undefined behavior, which isn't the case.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
I think the reason PG allows triggers to change the cascaded row is because PG still performs the FK checks at the end. Here's an example:

```sql
CREATE TABLE products (
    product_id INT PRIMARY KEY,
    product_name VARCHAR(100) NOT NULL
);

INSERT INTO products (product_id, product_name) VALUES
(1, 'Laptop'),
(2, 'Smartphone'),
(3, 'Tablet'),
(4, 'Smartwatch');

CREATE TABLE order_items (
    item_id INT PRIMARY KEY,
    product_id INT REFERENCES products(product_id) ON UPDATE CASCADE,
    quantity INT NOT NULL
);

INSERT INTO order_items (item_id, product_id, quantity) VALUES
(1, 1, 2),  -- 2 Laptops
(2, 2, 5),  -- 5 Smartphones
(3, 3, 3);  -- 3 Tablets

CREATE OR REPLACE FUNCTION adjust_product()
RETURNS TRIGGER AS $$
BEGIN
    IF NEW.product_id = 1 THEN
        NEW.product_id = 5;
    END IF;
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

CREATE TRIGGER product_update_trigger
BEFORE UPDATE ON order_items
FOR EACH ROW
WHEN (OLD.product_id IS DISTINCT FROM NEW.product_id)
EXECUTE FUNCTION adjust_product();

UPDATE order_items
SET product_id = 1
WHERE item_id = 2;
```

This will fail in PG with the following error:

```
ERROR:  insert or update on table "order_items" violates foreign key constraint "order_items_product_id_fkey"
DETAIL:  Key (product_id)=(5) is not present in table "products".
```

If 5 is inserted into products it will succeed.

Maybe instead of this check in the trigger, we should add the FK check when there is a trigger that could interfere with the FK cascade?

> Reviewed 10 of 10 files at r6, 1 of 4 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-:-OAd12XnEX7YgM5uML8T:b-r8on4m)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball)



<!-- Sent from Reviewable.io -->

DrewKimball commented:
>Can you explain more about the motivation for this setting? I don't see why it is necessary - Postgres allows triggers to modify cascading updates by default.
>If we do want to keep it, I think the word "unsafe" is a bit too strong to use in this case - it sounds like it can lead to unexpected or undefined behavior, which isn't the case.

Here's a snippet from the Postgres docs:
```
There is also nonstandard behavior if BEFORE triggers modify rows or prevent updates
during an update that is caused by a referential action. This can lead to constraint
violations or stored data that does not honor the referential constraint.
```
In short, I don't want to have to worry about users corrupting their DB, at least in the first version of trigger support. We'd have the same problem, unless we did what Michael's suggesting and added FK checks for every FK (addressed below).

This is also why I went with "unsafe". I'm still open to convincing, but this seems pretty bad to me.

>Maybe instead of this check in the trigger, we should add the FK check when there is a trigger that could interfere with the FK cascade?

Hypothetically we could do this - just assume that all rows could be modified by the trigger and add checks accordingly. The difficulty is in patterns like the diamond cascade pattern - the tables are temporarily in an inconsistent state between when one child table cascades and the other does. Running the FK check for the second child after the cascade for the first child would result in a FK violation error that shouldn't have happened, even if the trigger didn't actually modify any rows. So, I wanted to prefer the common case when the trigger doesn't modify the row.

Your earlier idea to allow modifications to non-FK columns seems safe to me, though. But I'd prefer to keep the changes minimal and the behavior more restrictive for now, and then we can potentially relax it later.

mgartner commented:
Notes from a conversation we had about this:

- Let's go ahead and merge this for 24.3 to prevent any chance of "FK corruption" by default.
- Ideally in a future release we can find a way to always perform an FK check after triggers fire to avoid FK corruption.

@michae2 @DrewKimball Feel free to add to this if I missed any key points.

michae2 reviewed (commented):
Nice job fixing those problems with diamond cascades. This is tricky stuff.

I do think we should be preventing FK constraint violations with FK checks, rather than this `ensureNoRowsModifiedByTrigger` check, but this is fine for 24.3.

Can there be AFTER queries on cascaded-to tables?

> Reviewed 3 of 4 files at r7, 1 of 1 files at r8, 3 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-:-OAdfAJc3M6tMR--AojU:b-wssphc)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball)

___
*[`pkg/sql/opt/optbuilder/trigger.go` line 298 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OAdXfejE3a96FgVgV-p:-OAcwn8lBETJwr8cO_A2:biq5zwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbd8f79ba5dead494f2fee4cea1931511af9096/pkg/sql/opt/optbuilder/trigger.go#L298)):*
> ```Go
> 	// Construct a call to crdb_internal.plpgsql_raise with the error message.
> 	message := fmt.Sprintf(
> 		"trigger %s attempted to modify or filter a mutated row in a cascade operation", triggerName,
> ```

It can be difficult to debug errors like this without knowing which row is affected. Would it be possible to include the original row values in the message?
___
*[`pkg/sql/opt/optbuilder/trigger.go` line 123 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OAdkYv75WvBuxBL_4sQ:-OAdkYv75WvBuxBL_4sR:bcxnir0) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfe57f46d08ef8f90082ef2c3905d09e39059a3e/pkg/sql/opt/optbuilder/trigger.go#L123)):*
> ```Go
> 				triggerScope, triggers[i].Name(), eventType, triggerFnColID, oldColID, newColID,
> 			)
> 			canModifyRows = false
> ```

Shouldn't `canModifyRows` be true if `UnsafeAllowTriggersModifyingCascades` is true? Otherwise I think we will lose the changes made by a trigger?

Maybe just lift the `UnsafeAllowTriggersModifyingCascades` check out of `ensureNoRowsModifiedByTrigger` to here?
___
*[`pkg/sql/opt/optbuilder/testdata/trigger` line 59 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OAdZdAJ2L03sytmgu36:-OAdZdAJ2L03sytmgu37:bq3e45m) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbd8f79ba5dead494f2fee4cea1931511af9096/pkg/sql/opt/optbuilder/testdata/trigger#L59)):*
> ```
> 
> build-post-queries format=(hide-all,show-columns)
> UPDATE xy SET y = 3 WHERE x = 1;
> ```

Why does this have to cascade? Because the BEFORE trigger on xy could change x?
___
*[`pkg/sql/opt/optbuilder/testdata/trigger` line 160 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OAd_zKR2zj4ODib5Bbz:-OAd_zKR2zj4ODib5Bc-:bvfpz89) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbd8f79ba5dead494f2fee4cea1931511af9096/pkg/sql/opt/optbuilder/testdata/trigger#L160)):*
> ```
>            │         ├── (f:53).k [as=k_new:55]
>            │         └── (f:53).x [as=x_new:56]
>            └── f-k-checks
> ```

Just noting that this seems like a case where we don't need the "trigger attempted to modify a mutated row" check, since we've already got the FK check afterward. But I guess cascaded deletes don't have the FK check.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
Thanks for the reviews!

>I do think we should be preventing FK constraint violations with FK checks, rather than this `ensureNoRowsModifiedByTrigger` check, but this is fine for 24.3.

I'll open an issue.

>Can there be AFTER queries on cascaded-to tables?

Yes, for example, in the diamond test case there are BEFORE+AFTER triggers on the child and grandchild tables. Is that what you meant?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-:-OAsbO7BBPGrj-Df6iQO:b-lzqcbn)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2)

___
*[`pkg/sql/opt/optbuilder/trigger.go` line 298 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OAdXfejE3a96FgVgV-p:-OAtxEBR89sBIP2xdkxh:b-lhaqqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbd8f79ba5dead494f2fee4cea1931511af9096/pkg/sql/opt/optbuilder/trigger.go#L298)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

It can be difficult to debug errors like this without knowing which row is affected. Would it be possible to include the original row values in the message?
</blockquote></details>

Yes - I've updated the message to include the expected values for the row.
___
*[`pkg/sql/opt/optbuilder/trigger.go` line 123 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OAdkYv75WvBuxBL_4sQ:-OAsc9XdA1NcNJzHXaef:b-tk2o3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfe57f46d08ef8f90082ef2c3905d09e39059a3e/pkg/sql/opt/optbuilder/trigger.go#L123)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Shouldn't `canModifyRows` be true if `UnsafeAllowTriggersModifyingCascades` is true? Otherwise I think we will lose the changes made by a trigger?

Maybe just lift the `UnsafeAllowTriggersModifyingCascades` check out of `ensureNoRowsModifiedByTrigger` to here?
</blockquote></details>

Great catch, done.
___
*[`pkg/sql/opt/optbuilder/testdata/trigger` line 59 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OAdZdAJ2L03sytmgu36:-OAu4va8Cv7qJ3oceEcM:b-6rh2z0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbd8f79ba5dead494f2fee4cea1931511af9096/pkg/sql/opt/optbuilder/testdata/trigger#L59)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Why does this have to cascade? Because the BEFORE trigger on xy could change x?
</blockquote></details>

Yes, that's exactly it. And if the BEFORE trigger doesn't modify any rows, the cascade will have no effect because of the `x_old:36 IS DISTINCT FROM x_new:37` filter.
___
*[`pkg/sql/opt/optbuilder/testdata/trigger` line 160 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OAd_zKR2zj4ODib5Bbz:-OAu5uh88WlSTB45fLRL:b-uc2aex) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbd8f79ba5dead494f2fee4cea1931511af9096/pkg/sql/opt/optbuilder/testdata/trigger#L160)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Just noting that this seems like a case where we don't need the "trigger attempted to modify a mutated row" check, since we've already got the FK check afterward. But I guess cascaded deletes don't have the FK check.
</blockquote></details>

It's subtle (and I forgot about this when we discussed earlier), but that FK check actually fires on the input of the cascaded mutation, not on the original mutation. So, let's say the trigger did filter some rows - they would be excluded from the check.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
Issue for relaxing/removing the runtime checks: https://github.com/cockroachdb/cockroach/issues/134285

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

> Reviewed 7 of 7 files at r10, 4 of 4 files at r11, 1 of 1 files at r12, 3 of 3 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-:-OAyzwVl8omHovizeO-r:bcmw7ss)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball)

___
*[`pkg/sql/opt/optbuilder/trigger.go` line 298 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OAdXfejE3a96FgVgV-p:-OAyvSxtDnrVqpiUtmbj:b-9755fn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbd8f79ba5dead494f2fee4cea1931511af9096/pkg/sql/opt/optbuilder/trigger.go#L298)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Yes - I've updated the message to include the expected values for the row.
</blockquote></details>

Thank you!
___
*[`pkg/sql/opt/optbuilder/testdata/trigger` line 160 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OAd_zKR2zj4ODib5Bbz:-OAyxeqdCd-7z_f14POo:bhbod7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbd8f79ba5dead494f2fee4cea1931511af9096/pkg/sql/opt/optbuilder/testdata/trigger#L160)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

It's subtle (and I forgot about this when we discussed earlier), but that FK check actually fires on the input of the cascaded mutation, not on the original mutation. So, let's say the trigger did filter some rows - they would be excluded from the check.
</blockquote></details>

Isn't this exactly what we want?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133220)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2)

___
*[`pkg/sql/opt/optbuilder/testdata/trigger` line 160 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OAd_zKR2zj4ODib5Bbz:-OAz1mfDEvEfi4Mtjvlj:b-7lt7ec) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbd8f79ba5dead494f2fee4cea1931511af9096/pkg/sql/opt/optbuilder/testdata/trigger#L160)):*
>Isn't this exactly what we want?

The cascade is supposed to make sure all rows in the child table are consistent with those of the parent table. If the BEFORE trigger filters some of those child rows in the cascade, then all those filtered rows will be left referencing the old value, leaving the DB in an inconsistent state.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133220)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball)

___
*[`pkg/sql/opt/optbuilder/testdata/trigger` line 160 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/133220#-OAd_zKR2zj4ODib5Bbz:-OB1W8LJBZe2v8BLJlt_:bkdkwz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffbd8f79ba5dead494f2fee4cea1931511af9096/pkg/sql/opt/optbuilder/testdata/trigger#L160)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

>Isn't this exactly what we want?

The cascade is supposed to make sure all rows in the child table are consistent with those of the parent table. If the BEFORE trigger filters some of those child rows in the cascade, then all those filtered rows will be left referencing the old value, leaving the DB in an inconsistent state.
</blockquote></details>

Ahhh, great point. Thank you.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
TFTRs!

bors r+
